File: app/src/main/java/com/shevaalex/android/rickmortydatabase/source/network/net_utils/InitManager.java
package com.shevaalex.android.rickmortydatabase.source.network.net_utils;

import android.util.Log;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import com.shevaalex.android.rickmortydatabase.models.ApiObjectModel;
import com.shevaalex.android.rickmortydatabase.models.ApiPageInfoModel;
import com.shevaalex.android.rickmortydatabase.models.ApiPageModel;
import com.shevaalex.android.rickmortydatabase.source.network.ApiResponse;

import java.util.ArrayList;
import java.util.List;

// ObjectModel: Single object type of a given category
// PageModelObject: Type for the API response (API response with category page data)
public abstract class InitManager<ObjectModel extends ApiObjectModel, PageModelObject extends ApiPageModel> {
    private static final String TAG = "LOG_TAG_InitManager";
    //sets the time in days when the refresh is due
    private static final int REFRESH_CONSTANT = 30;
    //result observable object to get status of network and database
    private MediatorLiveData<Resource<ObjectModel>> resultObjectModel = new MediatorLiveData<>();

    public InitManager() {
        init();
    }

    private void init(){
        //set the loading status for livedata
        resultObjectModel.setValue(Resource.loading(null));
        //observe last object livedata from the database
        final LiveData<ObjectModel> dbSourceLastObject = getLastEntryFromDb();
        final LiveData<ApiResponse<PageModelObject>> firstPage = createCall();
        resultObjectModel.addSource(dbSourceLastObject, lastObjectFromDb -> {
            if (lastObjectFromDb != null) {
                Log.d(TAG, "init: last db object ID: " + lastObjectFromDb.getId());
            }
            //call and observe the first api page
            resultObjectModel.addSource(firstPage, pageModelObjectApiResponse -> {
                //if response is successfull -> get the last object's id and fetch it
                if (pageModelObjectApiResponse instanceof ApiResponse.SuccessApiResponse) {
                    ApiPageInfoModel pageInfo =
                            ((ApiResponse.SuccessApiResponse<PageModelObject>) pageModelObjectApiResponse)
                                    .getBody()
                                    .getApiPageInfoModel();
                    int lastModelId = pageInfo.getCount();
                    int pageCount = pageInfo.getPages();
                    LiveData<ApiResponse<ObjectModel>> apiSourcelastObject
                            = callLastApiModel(lastModelId);
                    // if ApiResponse with last object is successfull - set it as Resource.success value
                    resultObjectModel.addSource(apiSourcelastObject, lastNetworkObject -> {
                        Log.d(TAG, "init: success ApiResponse single object");
                        if (lastNetworkObject instanceof ApiResponse.SuccessApiResponse) {
                            resultObjectModel
                                    .setValue(Resource
                                    .success(((ApiResponse.SuccessApiResponse<ObjectModel>) lastNetworkObject)
                                            .getBody()));
                            shouldFetch(((ApiResponse.SuccessApiResponse<ObjectModel>) lastNetworkObject)
                                            .getBody()
                                    , lastObjectFromDb, pageCount);
                        } else {
                            manageEmptyOrErrorResponse(lastNetworkObject);
                        }
                        resultObjectModel.removeSource(apiSourcelastObject);
                    });
                } else {
                    manageEmptyOrErrorResponse(pageModelObjectApiResponse);
                }
                resultObjectModel.removeSource(firstPage);
            });
            resultObjectModel.removeSource(dbSourceLastObject);
        });
    }

    /**
     * Called to decide whether an update of the database is needed
     * 1) Fetch if no cached object found
     * 2) Fetch if objects are different (eg language changed)
     * 3) Fetch if timestamp is older than 30 days
     * 4) Fetch if last network object id > database objects count
     */
    private void shouldFetch(ObjectModel lastNetworkObject,
                             ObjectModel lastCacheObject,
                             int pageCount) {
        final LiveData<Integer> dbObjectsCount = getDbEntriesCount();
        //fetch if no cached object found
        if (lastCacheObject == null) {
            Log.d(TAG, "shouldFetch true: last cache obj is null");
            fetchFromNetwork(pageCount);
            return;
        }
        //fetch if objects are not equal
        if (!areNetworkAndDbEntriesEqual(lastNetworkObject, lastCacheObject)) {
            Log.d(TAG, "shouldFetch true: objects are not equal");
            fetchFromNetwork(pageCount);
            return;
        }
        //fetch if timestamp is older than 30 days
        if (isTimestampObsolete(lastCacheObject)) {
            Log.d(TAG, "shouldFetch true: timestamp is older than refreshConstant");
            fetchFromNetwork(pageCount);
            return;
        }
        //observe number of saved entries in the database
        resultObjectModel.addSource(dbObjectsCount, dbCount -> {
            if (dbCount == null) {
                dbCount = 0;
            }
            Log.d(TAG, "shouldFetch: lastNetworkObject.getId() + dbCount: "
                    + lastNetworkObject.getId() + " + " + dbCount);
            //fetch if last network object id > number of database entries
            if (lastNetworkObject.getId() > dbCount) {
                fetchFromNetwork(pageCount);
            } else {
                Log.d(TAG, "shouldFetch: fetch not needed");
            }
            resultObjectModel.removeSource(dbObjectsCount);
        });
    }

    private void fetchFromNetwork(int pageCount){
        Log.d(TAG, "fetchFromNetwork: getting new data...");
        LiveData<List<ApiResponse<PageModelObject>>> pageObjectsList = callAllPages(pageCount);
        ArrayList<ApiResponse.SuccessApiResponse<PageModelObject>> apiResponsePageList
                = new ArrayList<>();
        resultObjectModel.addSource(pageObjectsList, apiResponseList -> {
            if (apiResponseList != null && !apiResponseList.isEmpty()) {
                Log.d(TAG, "fetchFromNetwork: apiResponseList.size= " + apiResponseList.size());
                for (ApiResponse<PageModelObject> singleApiResponsePage : apiResponseList) {
                    if (singleApiResponsePage instanceof ApiResponse.SuccessApiResponse) {
                        Log.d(TAG, "fetchFromNetwork: singleApiResponsePage= "
                                + ((ApiResponse.SuccessApiResponse<PageModelObject>) singleApiResponsePage).getBody().getApiPageInfoModel().toString());
                        if (!apiResponsePageList.contains(singleApiResponsePage)) {
                            apiResponsePageList
                                    .add((ApiResponse.SuccessApiResponse<PageModelObject>) singleApiResponsePage);
                        }
                    } else {
                        onFetchFailed();
                    }
                }
                Log.d(TAG, "fetchFromNetwork: apiResponsePageList.size= " + apiResponsePageList.size());
                //save result to database when all data is available
                if (apiResponsePageList.size() == pageCount) {
                    saveCallResult(apiResponsePageList);
                    resultObjectModel.removeSource(pageObjectsList);
                }
                if (apiResponseList.get(0) instanceof ApiResponse.EmptyApiResponse
                        || apiResponseList.get(0) instanceof ApiResponse.ErrorApiResponse) {
                    manageEmptyOrErrorResponse(apiResponseList.get(0));
                }
            }
        });
    }

    // if response is empty or error -> set the value as Resource.error
    private <T> void manageEmptyOrErrorResponse(ApiResponse<T> notSuccessfullResponse) {
        if (notSuccessfullResponse instanceof ApiResponse.EmptyApiResponse) {
            Log.d(TAG, "manageEmptyOrErrorResponse: EmptyApiResponse");
            resultObjectModel
                    .setValue(Resource.error("EmptyApiResponse", null));
        } else if (notSuccessfullResponse instanceof ApiResponse.ErrorApiResponse) {
            Log.d(TAG, "manageEmptyOrErrorResponse: ErrorApiResponse");
            resultObjectModel
                    .setValue(Resource
                            .error(((ApiResponse.ErrorApiResponse<T>) notSuccessfullResponse)
                                    .getErrorMessage(), null));
        }
        if (notSuccessfullResponse == null) {
            Log.d(TAG, "manageEmptyOrErrorResponse: EmptyApiResponse");
            resultObjectModel
                    .setValue(Resource.error("null object in response", null));
        }
    }

    //compares network and database objects
    private boolean areNetworkAndDbEntriesEqual(ObjectModel networkObjectModel,
                                                ObjectModel cacheObjectModel) {
        if (cacheObjectModel != null) {
            Log.d(TAG, "areNetworkAndDbEntriesEqual: " + networkObjectModel.equals(cacheObjectModel));
            Log.d(TAG, "areNetworkAndDbEntriesEqual: net/cache= "
                    + networkObjectModel.getId() + " / " + cacheObjectModel.getId());
        }
        return networkObjectModel.equals(cacheObjectModel);
    }

    /**
     * gets current time in days and compares it to a timestamp from the cache object
     * @return true if timestamp is older than REFRESH_CONSTANT
     */
    private boolean isTimestampObsolete(ObjectModel lastCacheObject) {
        int currentTime = (int) (System.currentTimeMillis()/86400000);
        Log.d(TAG, "shouldFetch: currentTime= " + currentTime);
        int lastRefresh = lastCacheObject.getTimeStamp();
        Log.d(TAG, "shouldFetch: lastRefresh= " + lastRefresh);
        Log.d(TAG, "shouldFetch: last refreshed: "
                + (currentTime - lastRefresh)
                + " days ago");
        return (currentTime - lastRefresh) >= REFRESH_CONSTANT;
    }

    // Called when the fetch fails.
    protected void onFetchFailed() {
        Log.d(TAG, "onFetchFailed: fetch failed for at least one ApiResponse<PageModelObject>");
        resultObjectModel
                .setValue(Resource.error("Fetch failed", null));
    }

    @WorkerThread
    protected abstract LiveData<ApiResponse<ObjectModel>> callLastApiModel(int lastModelId);

    //returns last entry from database
    @NonNull @MainThread
    protected abstract LiveData<ObjectModel> getLastEntryFromDb();

    //returns number of entries from database
    @NonNull @MainThread
    protected abstract LiveData<Integer> getDbEntriesCount();

    //calls all pages in sequence and returns a merged result
    @NonNull
    @MainThread
    protected abstract LiveData<List<ApiResponse<PageModelObject>>> callAllPages(int pageCount);

    //calls first page of a given category
    @NonNull
    @MainThread
    protected abstract LiveData<ApiResponse<PageModelObject>> createCall();

    // Called to save the result of the API response into the database
    @WorkerThread
    protected abstract void saveCallResult(List<ApiResponse.SuccessApiResponse<PageModelObject>> apiResponseList);

    // Returns a LiveData object that represents the resource that's implemented
    // in the base class. Used as a status object to update the state of the UI (eg loading / sync needed)
    public final LiveData<Resource<ObjectModel>> getAsLiveData() {
        return resultObjectModel;
    }

}

File: app/src/main/java/com/shevaalex/android/rickmortydatabase/repository/InitManager.kt
package com.shevaalex.android.rickmortydatabase.repository

import android.content.Context
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import com.shevaalex.android.rickmortydatabase.models.ApiObjectModel
import com.shevaalex.android.rickmortydatabase.models.ApiPageModel
import com.shevaalex.android.rickmortydatabase.utils.Constants
import com.shevaalex.android.rickmortydatabase.utils.UiTranslateUtils
import com.shevaalex.android.rickmortydatabase.utils.networking.ApiResult
import com.shevaalex.android.rickmortydatabase.utils.networking.Message
import com.shevaalex.android.rickmortydatabase.utils.networking.StateResource
import com.shevaalex.android.rickmortydatabase.utils.networking.Status
import timber.log.Timber
import java.util.*

abstract class InitManager<ObjectModel: ApiObjectModel, PageModelObject: ApiPageModel>(
        val context: Context
) {

    /**
     * Returns a StateResource object representing a status of database (eg error / success)
     */
    suspend fun getDbState(): StateResource {
        //call the first api page
        when (val apiSourceFirstPage = createCall()) {
            //if response is successfull -> get the last object's id and fetch it
            is ApiResult.Success<PageModelObject> -> {
                apiSourceFirstPage.data.apiPageInfoModel.let { pageInfoModel ->
                    val apiSourcelastObject = callLastApiModel(pageInfoModel.count)
                    // if ApiResponse with last object is successfull - check if shouldFetch
                    if (apiSourcelastObject is ApiResult.Success<ObjectModel>) {
                        val lastNetworkObject = apiSourcelastObject.data
                        val lastCacheObject = getLastEntryFromDb()
                        val dbObjectsCount = getDbEntriesCount()
                        //translate if needed
                        if (Locale.getDefault().language.startsWith("ru")
                                || Locale.getDefault().language.startsWith("uk")) {
                            UiTranslateUtils.getTranslatedObject(context, lastNetworkObject)
                        }
                        Timber.d("asFlow: success ApiResponse last object= %s",
                                apiSourcelastObject.data.name)
                        return if (shouldFetch(lastCacheObject, lastNetworkObject, dbObjectsCount)) {
                            //fetch needed = loading, syncing database
                            fetchFromNetwork(pageInfoModel.pages)
                        } else {
                            //if fetch not needed = database is up to date
                            StateResource(Status.Success, Message.DbIsUpToDate)
                        }
                    } else {
                        return manageEmptyOrErrorResponse(apiSourcelastObject)
                    }
                }
            }
            else -> return manageEmptyOrErrorResponse(apiSourceFirstPage)
        }
    }

    /**
     * Called to decide whether an update of the database is needed
     * 1) Fetch if no cached object found
     * 2) Fetch if objects are different (eg language changed, object's updated etc)
     * 3) Fetch if timestamp is older than {Constants.REFRESH_CONSTANT} days
     * 4) Fetch if last network object id > database objects count
     */
    private fun shouldFetch(
            lastCacheObject: ObjectModel?,
            lastNetworkObject: ObjectModel,
            dbObjectsCount: Int): Boolean {
        //fetch if no cached object found
        if (lastCacheObject == null) {
            Timber.d("shouldFetch true: last cache obj is null")
            return true
        }
        //fetch if objects are not equal
        if (!areNetworkAndDbEntriesEqual(lastNetworkObject, lastCacheObject)) {
            Timber.d("shouldFetch true: objects are not equal: \n%s / %s",
                    "lastCacheObject name [${lastCacheObject.name}]",
                    "lastNetworkObject name [${lastNetworkObject.name}]")
            return true
        }
        //fetch if timestamp is older than {Constants.REFRESH_CONSTANT} days
        if (isTimestampObsolete(lastCacheObject)) {
            Timber.d("shouldFetch true: timestamp is older than refreshConstant")
            return true
        }
        //fetch if last network object id > number of database entries
        Timber.d("shouldFetch: lastNetworkObject.getId()= %d, dbCount= %d",
                lastNetworkObject.id,
                dbObjectsCount)
        return if (lastNetworkObject.id > dbObjectsCount) {
            true
        } else {
            Timber.i("shouldFetch: fetch not needed")
            false
        }
    }

    private suspend fun fetchFromNetwork(pageCount: Int): StateResource {
        Timber.i("fetchFromNetwork: getting new data...")
        val resultPageObjects: List<ApiResult<PageModelObject>> = callAllPages(pageCount)
        Timber.d("fetchFromNetwork: pageObjectsList.size= ${resultPageObjects.size}")
        val unsuccessfullResult = {
            resultPage: ApiResult<PageModelObject> ->
            resultPage !is ApiResult.Success<PageModelObject>
        }
        //filter list to successfull results only and save them to db
        val successResultPageObjects = resultPageObjects
                .filterIsInstance<ApiResult.Success<PageModelObject>>()
        /*if list with success results contains data for all pages
                            -> set Resource.success, save results to db*/
        return if (successResultPageObjects.size == pageCount
                && !resultPageObjects.any(unsuccessfullResult)) {
            val successPages = successResultPageObjects.map{it.data}
            saveCallResult(successPages)
            //set status to success -> database is up to date now
            StateResource(
                    Status.Success,
                    Message.DbIsUpToDate
            )
        }
        //or if list has unsuccessfull results set Resource.error
        else {
            Timber.e("fetchFromNetwork: fetch failed")
            val failedResult = resultPageObjects.find(unsuccessfullResult)
            return manageEmptyOrErrorResponse(failedResult)
        }
    }

    /**
     * manages error, failed or empty responses
     * @return StateResource.Error
     */
    private fun <T> manageEmptyOrErrorResponse(
            notSuccessfullResponse: ApiResult<T>?
    ): StateResource =
            when (notSuccessfullResponse) {
                is ApiResult.Failure -> StateResource(
                        status = Status.Error,
                        message = Message.ServerError(notSuccessfullResponse.statusCode?:0)
                )
                ApiResult.NetworkError -> StateResource(
                        status = Status.Error,
                        message = Message.NetworkError
                )
                ApiResult.Empty -> StateResource(
                        status = Status.Error,
                        message = Message.EmptyResponse
                )
                else -> StateResource(
                        status = Status.Error,
                        message = Message.ServerError(0)
                )
            }

    /**
     * compares network and database objects to define if fetch needed
     */
    private fun areNetworkAndDbEntriesEqual(networkObjectModel: ObjectModel,
                                                 cacheObjectModel: ObjectModel): Boolean {
        Timber.d("areNetworkAndDbEntriesEqual: [%b] + IDs: cache [%d], network [%d]",
                networkObjectModel == cacheObjectModel,
                cacheObjectModel.id,
                networkObjectModel.id)
        return networkObjectModel == cacheObjectModel
    }

    /**
     * gets current time in days and compares it to a timestamp from the cache object
     * @return true if timestamp is older than REFRESH_CONSTANT (days)
     */
    private fun isTimestampObsolete(lastCacheObject: ObjectModel): Boolean {
        val currentTimeDays = (System.currentTimeMillis() / 86400000).toInt()
        val lastRefresh = lastCacheObject.timeStamp
        Timber.d("shouldFetch: {currentTime=[%d]} {lastRefresh=[%d]}, last refreshed [%d] days ago",
                currentTimeDays,
                lastRefresh,
                currentTimeDays - lastRefresh)
        return currentTimeDays - lastRefresh >= Constants.DB_REFETCH_PERIOD
    }

    /**
     * call and return last entry from the server
     */
    protected abstract suspend fun callLastApiModel(lastModelId: Int): ApiResult<ObjectModel>

    /**
     * return last entry from the database
     */
    @MainThread
    protected abstract suspend fun getLastEntryFromDb(): ObjectModel?

    /**
     * return number of entries from the database
     */
    @MainThread
    protected abstract suspend fun getDbEntriesCount(): Int

    /**
     * cocurrently call all pages and map the results using flow
     * @return list of all api results with page models
     */
    @MainThread
    protected abstract suspend fun callAllPages(pageCount: Int): List<ApiResult<PageModelObject>>

    /**
     * call server's first page of a given category
     */
    @MainThread
    protected abstract suspend fun createCall(): ApiResult<PageModelObject>

    /**
     * saves the results of the API response into the database
     */
    @WorkerThread
    protected abstract suspend fun saveCallResult(pageModels: List<PageModelObject>)

}